### PR TITLE
Delete history when shredding site data

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/39115

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
See issue for STR

### CODE REVIEW

File: ios/browser/api/history/brave_history_api.mm [hunk 1]
```diff
@@ -109,6 +109,127 @@ - (NSDate*)dateAdded {
...
+    }
+    if (endDate) {
+      [self setEndDate:endDate];
+    }
+  }
+  return self;
+}
+
+- (void)setMaxCount:(NSUInteger)maxCount {
+  max_count_ = static_cast<int>(maxCount);
```
Comment (Brandon-T) [inline ios/browser/api/history/brave_history_api.mm:163]: Should not do this. It should be setting the property and get rid of the `max_count_` variable.

File: ios/browser/api/history/brave_history_api.mm [hunk 2]
```diff
@@ -109,6 +109,127 @@ - (NSDate*)dateAdded {
...
+}
+
+- (NSUInteger)maxCount {
+  return static_cast<NSUInteger>(max_count_);
+}
+
+- (void)setHostOnly:(BOOL)hostOnly {
+  host_only_ = hostOnly;
```
Comment (Brandon-T) [inline ios/browser/api/history/brave_history_api.mm:171]: Should not have a `host_only_` variable. All of these setters can be removed.

File: ios/browser/api/history/brave_history_api.mm [hunk 3]
```diff
@@ -109,6 +109,127 @@ - (NSDate*)dateAdded {
...
+    case history::QueryOptions::DuplicateHandling::REMOVE_DUPLICATES_PER_DAY:
+      return HistoryDuplicateHandlingIOSRemovePerDay;
+    case history::QueryOptions::DuplicateHandling::KEEP_ALL_DUPLICATES:
+      return HistoryDuplicateHandlingIOSKeepAll;
+  }
+}
+
+- (history::QueryOptions)queryOptions {
+  history::QueryOptions options;
+  options.max_count = max_count_;
```
Comment (Brandon-T) [inline ios/browser/api/history/brave_history_api.mm:224]: These options should all be set directly from the Objective-C++ properties instead of having duplicate variables per property.

File: ios/brave-ios/Sources/Brave/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -859,13 +862,31 @@ class TabManager: NSObject {
...
     )
 
+    // Delete the history for forgotten websites
+    if let historyAPI = self.historyAPI {
+      let nodes = await historyAPI.search(
+        withQuery: etldP1,
+        options: HistorySearchOptions(
+          maxCount: 0,
+          hostOnly: false,
+          duplicateHandling: .keepAll,
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/TabManager.swift:880]: Why did you change this to `keepAll` instead of deleting duplicates?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: ios/browser/api/history/brave_history_api.mm [hunk 5]
```diff
@@ -109,6 +109,127 @@ - (NSDate*)dateAdded {
...
+- (void)setHostOnly:(BOOL)hostOnly {
+  host_only_ = hostOnly;
+}
+
+- (BOOL)hostOnly {
+  return host_only_;
+}
+
+- (void)setBeginDate:(NSDate*)beginDate {
+  begin_time_ = base::Time::FromNSDate(beginDate);
```
Comment (Brandon-T) [inline ios/browser/api/history/brave_history_api.mm:179]: Remove these custom setters and getters.

File: ios/browser/api/history/brave_history_api.mm [hunk 6]
```diff
@@ -109,6 +109,127 @@ - (NSDate*)dateAdded {
...
+- (HistoryDuplicateHandlingIOS)duplicateHandling {
+  switch (duplicate_policy_) {
+    case history::QueryOptions::DuplicateHandling::REMOVE_ALL_DUPLICATES:
+      return HistoryDuplicateHandlingIOSRemoveAll;
+    case history::QueryOptions::DuplicateHandling::REMOVE_DUPLICATES_PER_DAY:
+      return HistoryDuplicateHandlingIOSRemovePerDay;
+    case history::QueryOptions::DuplicateHandling::KEEP_ALL_DUPLICATES:
+      return HistoryDuplicateHandlingIOSKeepAll;
+  }
+}
```
Comment (Brandon-T) [inline ios/browser/api/history/brave_history_api.mm:220]: Remove these custom settings and getters. Change to conversion functions and use the conversion functions at the call site.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: ios/browser/api/history/brave_history_api.mm (hunk 1 above)
Comment (cuba) [reply ios/browser/api/history/brave_history_api.mm:163]: Must just need a rebase or a bad rebase.

File: ios/browser/api/history/brave_history_api.mm (hunk 2 above)
Comment (cuba) [reply ios/browser/api/history/brave_history_api.mm:171]: Again some rebase problems

File: ios/brave-ios/Sources/Brave/Frontend/Browser/TabManager.swift (hunk 4 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/TabManager.swift:880]: Same

Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/TabManager.swift:880]: I think delete duplicates is a bad name (unless I misunderstood). It doesn't actually delete them. It basically doesn't return them. So a better name is perhaps skipDuplicates. In any case since I do want duplicates returned (as i'm deleting history) i need to pass keepAll to get duplicates so I can actually delete them too.

File: ios/browser/api/history/brave_history_api.mm (hunk 6 above)
Comment (cuba) [reply ios/browser/api/history/brave_history_api.mm]: I hope I got this right: https://github.com/brave/brave-core/pull/24286/commits/1e873881ad2389c828ed51c5cca1c010b9524fe8

**Review by Brandon-T — APPROVED** [review]
